pkg/proxy/ipvs/proxier.go

```diff
@@ -922,13 +926,26 @@ func (proxier *Proxier) syncProxyRules() {
 	// We assume that if this was called, we really want to sync them,
 	// even if nothing changed in the meantime. In other words, callers are
 	// responsible for detecting no-op changes and not calling this function.
+
+	// (khenidak) the previous assumption is valid only for single stack
+	// for dual stack, we have no way for endpointslice (or general endpoint processing)
+	// to know if this endpoint is being processed for an IPv6 or IPv4 proxier
+	// TODO: Find a way where the endpoints/slice can be loaded with knowledge
+	// of proxy/ipfamily and perform the filtering there.
 	serviceUpdateResult := proxy.UpdateServiceMap(proxier.serviceMap, proxier.serviceChanges)
 	endpointUpdateResult := proxier.endpointsMap.Update(proxier.endpointsChanges)
 
 	staleServices := serviceUpdateResult.UDPStaleClusterIP
 	// merge stale services gathered from updateEndpointsMap
 	for _, svcPortName := range endpointUpdateResult.StaleServiceNames {
 		if svcInfo, ok := proxier.serviceMap[svcPortName]; ok && svcInfo != nil && svcInfo.Protocol() == v1.ProtocolUDP {
+			// ignore stale services that does not match ipfamily of the proxier
+			isIpv6Svc := utilnet.IsIPv6(svcInfo.ClusterIP())
+			if proxier.isIPv6 != isIpv6Svc {
+				klog.V(10).Infof("ipvs proxier (ipv6:%v) is ignoring  a stale service:%v (ipv6:%v) because it is not of same ipfamily", proxier.isIPv6, svcPortName, isIpv6Svc)
+				continue
+			}
+
```

Comment:
Is this something we could do in the caches instead of `syncProxyRules`? As I understand it, right now a change to a v4 address would be considered a change by the v6 proxier since both addresses would be stored in both caches. If on the other hand we filtered addresses before storing them in the cache, changes would be more accurate by the time they got here.

Reply:
I think this would help with what @andrewsykim is asking about. I think https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/endpoints.go#L129 and https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/endpoints.go#L175 would be the starting point for that.